### ISSUE TITLE
error: cannot find symbol COMMA_JOINER

### ISSUE THREAD

**Issue by chrisalice** [google/FreeBuilder#157]
I believe there is still an issue with a undesired *indirect* dependency on Guava in case a `java.util.Optional` is used.

The `example.Person` class in the issue description has multiple fields. This results in the toString() method being generated in the inner `example.Person_Builder.Value` class as follows:

```java
package example;

...

@Generated("org.inferred.freebuilder.processor.CodeGenerator")
abstract class Person_Builder {

  ...
  
  private static final class Value implements Person {
    ...
    
    @Override
    public String toString() {
      return "Person{"
          + COMMA_JOINER.join("name=" + name, (age != null ? "age=" + age : null)) // <------- Compilation fails here as the COMMA_JOINER static final field is not generated.
          + "}";
    }
    
    ...

  }
  ...

}
```

When I include the Guava dependency everything compiles fine, because the field COMMA_JOINER gets generated, but it is however of type `com.google.common.base.Joiner` which belongs to the Guava dependency.

I'm not sure as I didn't debug the annotation processing, but the bug seems to be in the `org.inferred.freebuilder.processor.CodeGenerator#addValueType`:

```java
...

public class CodeGenerator {

  ...

  private static void addConstantDeclarations(Metadata metadata, SourceBuilder body) {
    if (body.feature(GUAVA).isAvailable() && metadata.getProperties().size() > 1) {
      body.addLine("")
          .addLine("private static final %1$s COMMA_JOINER = %1$s.on(\", \").skipNulls();",
              Joiner.class); // <------- This looks good, the COMMA_JOINER static final field only gets generated in case Guava is available.
    }
  }

  ...

  private static void addValueType(SourceBuilder code, Metadata metadata) {
    ...

    // toString
    if (metadata.standardMethodUnderride(StandardMethod.TO_STRING) == ABSENT) {
      code.addLine("")
          .addLine("  @%s", Override.class)
          .addLine("  public %s toString() {", String.class)
          .add("    return \"%s{", metadata.getType().getSimpleName());
      switch (metadata.getProperties().size()) {
        case 0: {
          code.add("}\";\n");
          break;
        }

        case 1: {
          Property property = getOnlyElement(metadata.getProperties());
          if (property.getCodeGenerator().getType() == Type.OPTIONAL) {
            code.add("\" + (%1$s != null ? \"%1$s=\" + %1$s : \"\") + \"}\";\n",
                property.getName()); // <------- When there is only one property and it is optional, everything is fine as there is no need to join anything with the COMMA_JOINER.
          } else {
            ...
          }
          break;
        }

        default: {
          // If one or more of the properties are optional, use COMMA_JOINER for readability.
          // Otherwise, use string concatenation for performance.
          if (any(metadata.getProperties(), IS_OPTIONAL)) {
            code.add("\"\n")
                .add("        + COMMA_JOINER.join(\n"); // <------- It looks as if there is more than one property and one or more of them are optional, COMMA_JOINER is always used whether or not it was generated. I.e. there is no 'Guava-safe' check.
            ...
          } else {
            ...
          }
          break;
        }
      }
      code.addLine("  }");
    }

  ...

}
```

So in short, the way the `toString()` method is generated for the value class is not completely working out in case someone has no dependency on Guava.

-- @dveijck

**Comment by chrisalice** [comment]
This should now be fixed in [1.10.3](https://github.com/google/FreeBuilder/releases/tag/v1.10.3).

**Comment by dveijck** [comment]
Hi @chrisalice, everything seems to compile just fine now with version 1.10.3 and without a Guava dependency. Thanks again for the quick response and fix. It is much appreciated!
